### PR TITLE
Crud resolver code generator based on entities

### PR DESCRIPTION
**Pull request by nsams**
crud code generation based on entity class

TODO:

- [x] support blocks in input: add missing transformToBlock
- [x] support Date fields in input
- [x] support Date fields in filter
- [x] support boolean fields in input
- [x] support boolean fields in filter
- [x] support number fields in input
- [x] support number fields in filter
- [x] filter: quoting for like
- [x] filter: support contains and startsWith/endsWith in one query
- [x] filter: move filter -> mikro-orm query code into library
- [x] support ignoring fields for input
- [x] new sort API
- [x] filter api naming: don't use shortcuts

TODO (later, not in this PR):
- [ ] evaluate alternative to lookupPath (for blocks): parse entity source file to find import paths
- [ ] support Enum fields in input
- [ ] support Enum fields in filter
- [ ] int vs. float (vs. decimal?)
- [ ] support relations?
- [ ] permissions? (different annotations)
- [ ] improve DX: currently after model change api has to be restarted to re-generate

### CODE REVIEW

File: packages/api/cms-api/src/generator/generate.ts [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+        const entity = entities[name];
+        {
+            const generatorOptions = Reflect.getMetadata(`data:crudGeneratorOptions`, entity.class) as CrudGeneratorOptions | undefined;
+            if (generatorOptions) {
+                await generateCrud(generatorOptions, entity);
+            }
+        }
+        {
+            const generatorOptions = Reflect.getMetadata(`data:crudSingleGeneratorOptions`, entity.class) as CrudSingleGeneratorOptions | undefined;
+            console.log(generatorOptions);
```
Comment (johnnyomair) [inline packages/api/cms-api/src/generator/generate.ts:23]:
```suggestion
```

**Review by johnnyomair — APPROVED** [review]
I think this is ready for use! 😁 👍🏼 

I like that you added tests for the generator as well.

File: packages/api/cms-api/src/generator/generate-crud.ts [hunk 2]
```diff
@@ -0,0 +1,387 @@
+import { EntityMetadata } from "@mikro-orm/core";
+import * as path from "path";
+
+import { CrudGeneratorOptions, hasFieldFeature } from "./crud-generator.decorator";
+import { writeCrudInput as writeCrudInput } from "./generate-crud-input";
```
Comment (johnnyomair) [inline packages/api/cms-api/src/generator/generate-crud.ts:5]:
```suggestion
import { writeCrudInput } from "./generate-crud-input";
```

**Review by johnnyomair — APPROVED** [review]

**Comment by nsams** [comment]
I'm unsure of naming and typing of functions in packages/api/cms-api/src/common/filter/mikro-orm.ts (I was not able to type this properly)

see also usage in generated code:
```
@Injectable()
export class ProductsService {
    getFindCondition(options: { query?: string; filter?: ProductFilter }): ObjectQuery<Product> {
        const andFilters = [];

        if (options.query) {
            andFilters.push(mikroOrmQueryFields(options.query, ["title", "slug"]));
        }

        if (options.filter) {
            andFilters.push(mikroOrmFilter(options.filter));
        }

        return andFilters.length > 0 ? { $and: andFilters } : {};
    }
}
``'

**Comment by johnnyomair** [comment]
### Naming
- `mikroOrmQueryFields` -> `queryToFilterQuery()` or `queryToMikroOrmQuery()`
- `mikroOrmFilter` -> `filtersToFilterQuery()` or `filtersToMikroOrmQuery()`

BTW, shouldn't `filterToMikroOrmQuery` be an internal API? Why do you expose it?

### Typing
I tried to improve typing by using Generics, but it's not simple. Is strict typing really necessary? As the code is only used in generated code?

**Comment by nsams** [comment]
> Naming

changed.

> I tried to improve typing by using Generics, but it's not simple. Is strict typing really necessary? As the code is only used in generated code?

well, if you use the code as scaffolding...

**Pull request merged**